sled-agent/src/bootstrap/server.rs

```diff
@@ -137,3 +639,312 @@ pub fn run_openapi() -> Result<(), String> {
 }
+
+struct Inner {
+    config: SledConfig,
+    hardware_monitor: broadcast::Receiver<HardwareUpdate>,
+    state: SledAgentState,
+    sled_init_rx: mpsc::Receiver<(
+        StartSledAgentRequest,
+        oneshot::Sender<Result<SledAgentResponse, String>>,
+    )>,
+    sled_reset_rx: mpsc::Receiver<oneshot::Sender<Result<(), BootstrapError>>>,
+    managers: BootstrapManagers,
+    ddm_admin_localhost_client: DdmAdminClient,
+    bootstore_handles: BootstoreHandles,
+    _sprockets_server_handle: JoinHandle<()>,
+    _key_manager_handle: JoinHandle<()>,
+    base_log: Logger,
+}
+
+impl Inner {
+    async fn run(mut self) {
+        let log = self.base_log.new(o!("component" => "SledAgentMain"));
+        loop {
+            // TODO-correctness We pause handling hardware update messages while
+            // we handle sled init/reset requests - is that okay?
+            tokio::select! {
+                // Cancel-safe per the docs on `broadcast::Receiver::recv()`.
+                hardware_update = self.hardware_monitor.recv() => {
+                    self.handle_hardware_update(hardware_update, &log).await;
+                }
+
+                // Cancel-safe per the docs on `mpsc::Receiver::recv()`.
+                Some((request, response_tx)) = self.sled_init_rx.recv() => {
+                    self.handle_start_sled_agent_request(
+                        request,
+                        response_tx,
+                        &log,
+                    ).await;
+                }
+
+                // Cancel-safe per the docs on `mpsc::Receiver::recv()`.
+                Some(response_tx) = self.sled_reset_rx.recv() => {
+                    // Try to reset the sled, but do not exit early on error.
+                    let result = async {
+                        self.uninstall_zones().await?;
+                        self.uninstall_sled_local_config().await?;
+                        self.uninstall_networking(&log).await?;
+                        self.uninstall_storage(&log).await?;
+                        Ok::<(), BootstrapError>(())
+                    }
+                    .await;
+
+                    _ = response_tx.send(result);
+                }
+            }
+        }
+    }
+
+    async fn handle_hardware_update(
+        &self,
+        hardware_update: Result<HardwareUpdate, broadcast::error::RecvError>,
+        log: &Logger,
+    ) {
+        info!(
+            log,
+            "Handling hardware update message";
+            "phase" => "bootstore-steady-state",
+            "update" => ?hardware_update,
+        );
+
+        // TODO are these choices correct? This tries to match the previous
+        // behavior when SledAgent monitored hardware itself. Could/should we
+        // notify nexus after any hardware update?
+        let should_notify_nexus = match &hardware_update {
+            Ok(HardwareUpdate::TofinoDeviceChange)
+            | Err(broadcast::error::RecvError::Lagged(_)) => true,
```

Comment:
We notified Nexus after these in particular because they could impact the `is_scrimlet` value that Nexus is tracking -- we identify sleds as scrimlets if the device is detected, even if the driver hasn't loaded.

(also, if we lagged, we could have missed any number of events, so that makes sense as a catch-all for "re-notify" IMO)

Arguably we could also tell Nexus information about what drivers have loaded, though we are definitely not doing that today.